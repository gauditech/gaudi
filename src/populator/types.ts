type PopulatorModel = {
  model: string;
  repeat: { min: number; max: number } | number;
  data: PopulatorData[];
};

type PopulatorData = {
  field: string;
  type: "constant" | "";
};
