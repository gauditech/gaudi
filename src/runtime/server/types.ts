import { NextFunction, Request, Response } from "express";

export type ServerHttpMethods = "all" | "get" | "post" | "put" | "patch" | "delete";

export type ServerRequestHandler =
  | ((request: Request, response: Response) => void | Promise<void>)
  | ((req: Request, resp: Response, next: NextFunction) => void | Promise<void>);

export type EndpointConfig = {
  path: string;
  method: ServerHttpMethods;
  handlers: ServerRequestHandler[];
};

export type ServerMiddlewareNextFn = (err?: unknown) => void;
