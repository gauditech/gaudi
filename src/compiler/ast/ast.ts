import { Type } from "./type";

export type ProjectASTs = {
  plugins: Record<string, GlobalAtom[]>;
  document: GlobalAtom[];
};

export type GlobalAtom = Model | Entrypoint | Populator | Runtime | Authenticator | Generator;

export type Model = {
  kind: "model";
  keyword: TokenData;
  name: Identifier;
  atoms: ModelAtom[];
};
export type ModelAtom = Field | Reference | Relation | Query | Computed | ModelHook;

export type Field = {
  kind: "field";
  keyword: TokenData;
  name: Identifier;
  ref: Ref;
  type: Type;
  atoms: FieldAtom[];
};
export type FieldAtom = { keyword: TokenData } & (
  | { kind: "type"; identifier: Identifier }
  | { kind: "unique" }
  | { kind: "nullable" }
  | { kind: "default"; literal: Literal }
  | { kind: "validate"; validators: Validator[] }
);
export type Validator =
  | FieldValidationHook
  | { kind: "builtin"; name: Identifier; args: Literal[] };

export type Reference = {
  kind: "reference";
  keyword: TokenData;
  name: Identifier;
  ref: Ref;
  type: Type;
  atoms: ReferenceAtom[];
};
export type ReferenceAtom = { keyword: TokenData } & (
  | { kind: "to"; identifier: IdentifierRef }
  | { kind: "nullable" }
  | { kind: "unique" }
);

export type Relation = {
  kind: "relation";
  keyword: TokenData;
  name: Identifier;
  ref: Ref;
  type: Type;
  atoms: RelationAtom[];
};
export type RelationAtom = { keyword: TokenData } & (
  | { kind: "from"; identifier: IdentifierRef }
  | { kind: "through"; identifier: IdentifierRef }
);

export type Query = {
  kind: "query";
  keyword: TokenData;
  name: Identifier;
  ref: Ref;
  type: Type;
  atoms: QueryAtom[];
};
export type QueryAtom = { keyword: TokenData } & (
  | {
      kind: "from";
      identifierPath: IdentifierRef[];
      as?: { keyword: TokenData; identifierPath: IdentifierRef[] };
    }
  | { kind: "filter"; expr: Expr<Db> }
  | { kind: "orderBy"; orderBy: OrderBy }
  | { kind: "limit"; value: IntegerLiteral }
  | { kind: "offset"; value: IntegerLiteral }
  | { kind: "select"; select: Select }
  | { kind: "aggregate"; aggregate: AggregateType }
);
export type AggregateType = "count" | "one" | "first";
export type OrderBy = (
  | {
      identifierPath: IdentifierRef[];
      keyword?: undefined;
      order?: undefined;
    }
  | {
      identifierPath: IdentifierRef[];
      order: OrderType;
      keyword: TokenData;
    }
)[];
export type OrderType = "asc" | "desc";

export type Computed = {
  kind: "computed";
  keyword: TokenData;
  name: Identifier;
  ref: Ref;
  type: Type;
  expr: Expr<Db>;
};

export type Entrypoint = {
  kind: "entrypoint";
  keyword: TokenData;
  target: IdentifierRef;
  atoms: EntrypointAtom[];
};

export type EntrypointAtom =
  | { kind: "response"; select: Select; keyword: TokenData }
  | { kind: "authorize"; expr: Expr<Code>; keyword: TokenData }
  | Endpoint
  | Entrypoint
  | Identify;

export type Identify = {
  kind: "identify";
  keyword: TokenData;
  as?: { keyword: TokenData; identifier: IdentifierRef };
  atoms: { kind: "through"; keyword: TokenData; identifier: IdentifierRef }[];
};

export type Endpoint = {
  kind: "endpoint";
  keyword: TokenData;
  keywordType: TokenData;
  type: EndpointType;
  atoms: EndpointAtom[];
};
export type EndpointType = "list" | "get" | "create" | "update" | "delete" | "custom";
export type EndpointMethod = "GET" | "POST" | "PATCH" | "DELETE";
export type EndpointCardinality = "one" | "many";

export type EndpointAtom = { keyword: TokenData } & (
  | { kind: "action"; actions: Action[] }
  | { kind: "authorize"; expr: Expr<Code> }
  | { kind: "method"; method: EndpointMethod; methodKeyword: TokenData }
  | { kind: "cardinality"; cardinality: EndpointCardinality; cardinalityKeyword: TokenData }
  | { kind: "path"; path: StringLiteral }
  | { kind: "pageable" }
  | { kind: "orderBy"; orderBy: OrderBy }
  | { kind: "filter"; expr: Expr<Db> }
);

export type Action = ModelAction | DeleteAction | ExecuteAction | FetchAction;

export type ModelAction = {
  kind: "create" | "update";
  keyword: TokenData;
  target?: IdentifierRef[];
  as?: { keyword: TokenData; identifier: IdentifierRef };
  atoms: ModelActionAtom[];
};
export type ModelActionAtom =
  | ActionAtomSet
  | ActionAtomReferenceThrough
  | ActionAtomDeny
  | ActionAtomInput
  | ActionAtomVirtualInput;

export type DeleteAction = {
  kind: "delete";
  keyword: TokenData;
  target?: IdentifierRef[];
};

export type ExecuteAction = {
  kind: "execute";
  keyword: TokenData;
  keywordAs?: TokenData;
  name?: Identifier;
  atoms: ExecuteActionAtom[];
};
export type ExecuteActionAtom =
  | ActionAtomVirtualInput
  | ActionHook
  | { kind: "responds"; keyword: TokenData };

export type FetchAction = {
  kind: "fetch";
  keyword: TokenData;
  keywordAs: TokenData;
  name: Identifier;
  atoms: FetchActionAtom[];
};
export type FetchActionAtom = ActionAtomVirtualInput | AnonymousQuery;

export type ActionAtomSet = {
  kind: "set";
  keyword: TokenData;
  target: IdentifierRef;
  set: ActionHook | { kind: "expr"; expr: Expr<Code> };
};
export type ActionAtomReferenceThrough = {
  kind: "referenceThrough";
  keyword: TokenData;
  target: IdentifierRef;
  through: IdentifierRef;
  keywordThrough: TokenData;
};
export type ActionAtomDeny = {
  kind: "deny";
  keyword: TokenData;
  fields:
    | { kind: "all"; keyword: TokenData }
    | {
        kind: "list";
        fields: IdentifierRef[];
      };
};
export type ActionAtomInput = {
  kind: "input";
  keyword: TokenData;
  fields: {
    field: IdentifierRef;
    atoms: InputAtom[];
  }[];
};
export type InputAtom = { keyword: TokenData } & (
  | { kind: "optional" }
  | { kind: "default"; value: Expr<Code> }
);
export type ActionAtomVirtualInput = {
  kind: "virtualInput";
  keyword: TokenData;
  name: Identifier;
  ref: Ref;
  type: Type;
  atoms: ActionAtomVirtualInputAtom[];
};

export type ActionAtomVirtualInputAtom = { keyword: TokenData } & (
  | { kind: "type"; identifier: Identifier }
  | { kind: "nullable" }
  | { kind: "validate"; validators: Validator[] }
);

export type Populator = {
  kind: "populator";
  keyword: TokenData;
  name: Identifier;
  atoms: Populate[];
};
export type Populate = {
  kind: "populate";
  keyword: TokenData;
  name: Identifier;
  atoms: PopulateAtom[];
};
export type PopulateAtom =
  | {
      kind: "target";
      keyword: TokenData;
      identifier: IdentifierRef;
      as?: { keyword: TokenData; identifier: IdentifierRef };
    }
  | {
      kind: "repeat";
      keyword: TokenData;
      as?: { keyword: TokenData; identifier: IdentifierRef };
      repeatValue: RepeatValue;
    }
  | ActionAtomSet
  | Populate;

export type RepeatValue =
  | { kind: "long"; atoms: RepeatAtom[] }
  | { kind: "short"; value: IntegerLiteral };
export type RepeatAtom = { keyword: TokenData } & (
  | { kind: "start"; value: IntegerLiteral }
  | { kind: "end"; value: IntegerLiteral }
);

export type GeneratorType = "client";
export type Generator = {
  kind: "generator";
  keyword: TokenData;
} & {
  type: Extract<GeneratorType, "client">;
  keywordType: TokenData;
  atoms: GeneratorClientAtom[];
};
export type GeneratorClientAtom =
  | {
      kind: "target";
      keyword: TokenData;
      value: GeneratorClientAtomTarget;
      keywordValue: TokenData;
    }
  | { kind: "api"; keyword: TokenData; value: GeneratorClientAtomApi; keywordValue: TokenData }
  | { kind: "output"; keyword: TokenData; value: StringLiteral };
export type GeneratorClientAtomTarget = "js";
export type GeneratorClientAtomApi = "entrypoint" | "model";

export type Runtime = {
  kind: "runtime";
  keyword: TokenData;
  name: Identifier;
  atoms: RuntimeAtom[];
};
export type RuntimeAtom = { keyword: TokenData } & (
  | { kind: "default" }
  | { kind: "sourcePath"; path: StringLiteral }
);

export type Authenticator = {
  kind: "authenticator";
  keyword: TokenData;
  atoms: AuthenticatorAtom[];
};
export type AuthenticatorAtom = { kind: "method"; keyword: TokenData; method: AuthenticatorMethod };
export type AuthenticatorMethod = { kind: "basic"; keyword: TokenData };

export type Hook<named extends boolean, simple extends boolean> = {
  kind: "hook";
  keyword: TokenData;
  name: named extends true ? Identifier : undefined;
  ref: named extends true ? Ref : undefined;
  atoms: (
    | (simple extends true
        ? { kind: "default_arg"; keyword: TokenData; name: Identifier }
        :
            | { kind: "arg_expr"; keyword: TokenData; name: Identifier; expr: Expr<Code> }
            | { kind: "arg_query"; keyword: TokenData; name: Identifier; query: AnonymousQuery })
    | {
        kind: "source";
        keyword: TokenData;
        keywordFrom: TokenData;
        name: Identifier;
        file: StringLiteral;
        runtimePath?: string;
      }
    | { kind: "inline"; keyword: TokenData; code: StringLiteral }
    | { kind: "runtime"; keyword: TokenData; identifier: Identifier }
  )[];
};
export type ModelHook = Hook<true, false> & { type: Type };
export type FieldValidationHook = Hook<false, true>;
export type ActionHook = Hook<false, false>;

export type AnonymousQuery = {
  kind: "anonymousQuery";
  keyword: TokenData;
  atoms: QueryAtom[];
  type: Type;
};

export type Select = {
  target:
    | { kind: "short"; name: IdentifierRef }
    | { kind: "long"; name: Identifier; identifierPath: IdentifierRef[] };
  select?: Select;
}[];

export type Db = "db";
export type Code = "code";
export type ExprKind = Db | Code;
export type Expr<kind extends ExprKind = ExprKind> = (
  | {
      kind: "binary";
      keyword: TokenData;
      operator: BinaryOperator;
      lhs: Expr<kind>;
      rhs: Expr<kind>;
    }
  | { kind: "group"; expr: Expr<kind> }
  | { kind: "unary"; keyword: TokenData; operator: UnaryOperator; expr: Expr<kind> }
  | { kind: "path"; path: IdentifierRef[] }
  | { kind: "literal"; literal: Literal }
  | { kind: "function"; name: Identifier; args: Expr<kind>[] }
) & { type: Type; sourcePos: TokenData };
export type BinaryOperator =
  | "or"
  | "and"
  | "is not"
  | "is"
  | "not in"
  | "in"
  | "<"
  | "<="
  | ">"
  | ">="
  | "+"
  | "-"
  | "/"
  | "*";
export type UnaryOperator = "not";

export type Literal = IntegerLiteral | FloatLiteral | BooleanLiteral | NullLiteral | StringLiteral;
export type IntegerLiteral = { kind: "integer"; value: number; token: TokenData };
export type FloatLiteral = { kind: "float"; value: number; token: TokenData };
export type BooleanLiteral = { kind: "boolean"; value: boolean; token: TokenData };
export type NullLiteral = { kind: "null"; value: null; token: TokenData };
export type StringLiteral = { kind: "string"; value: string; token: TokenData };

export type RefUnresolved = { kind: "unresolved" };
export type RefModel = { kind: "model"; model: string };
export type RefModelAtom = {
  kind: "modelAtom";
  atomKind: ModelAtom["kind"];
  name: string;
  model: string;
  unique: boolean;
};
export type ContextKind =
  | "entrypointTarget"
  | "populateTarget"
  | "fetch"
  | "virtualInput"
  | "repeat"
  | "authToken"
  | "struct";
export type RefContext = { kind: "context"; contextKind: ContextKind };
export type Ref = RefUnresolved | RefModel | RefModelAtom | RefContext;

export const unresolvedRef: Ref = { kind: "unresolved" };

export type Identifier = { text: string; token: TokenData };
export type IdentifierRef = { identifier: Identifier; ref: Ref; type: Type };

export type TokenData = { start: number; end: number };

export type Parsed = "parsed";
export type Resolved = "resolved";
export type Typed = "typed";
export type Stage = Parsed | Resolved | Typed;
