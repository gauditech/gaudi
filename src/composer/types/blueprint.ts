import { Model } from "./model";

export interface Blueprint {
  models: Record<string, Model>;
}
