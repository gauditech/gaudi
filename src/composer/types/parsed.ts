export interface ModelDef {
  name: string;
  fields: FieldDef[];
  references?: ReferenceDef[];
  relations?: RelationDef[];
}

export interface FieldDef {
  name: string;
  type: "string" | "integer" | "datetime";
  nullable?: boolean;
}

export interface ReferenceDef {
  name: string;
  model: string;
}

export interface RelationDef {
  name: string;
  model: string;
}
