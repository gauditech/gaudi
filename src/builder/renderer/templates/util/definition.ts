import { Definition, PathFragment } from "@src/types/definition";

/*
 * Template helper functions for Gaudi definition
 */

export function modelDbName(modelRefKey: string, definition: Definition) {
  return (definition.models || []).find((model) => model.refKey === modelRefKey)?.dbname;
}

export function fieldDbName(modelRefKey: string, fieldRefKey: string, definition: Definition) {
  return (definition.models || [])
    .find((model) => model.refKey === modelRefKey)
    ?.fields.find((field) => field.refKey === fieldRefKey)?.dbname;
}

export function typeToDbType(type: string) {
  if (type === "serial") {
    return "Int";
  } else if (type === "integer") {
    return "Int";
  } else if (type === "text") {
    return "String";
  } else if (type === "boolean") {
    return "Boolean";
  }

  throw "UNHANDLED_GAUDI_TYPE";
}

export function buildEndpointPath(parts: PathFragment[]): string {
  return parts.reduce((accum, part) => {
    const pathPart =
      part.type === "numeric" || part.type === "text" ? `:${part.varname}` : part.value;
    return `${accum}/${pathPart}`;
  }, "");
}
