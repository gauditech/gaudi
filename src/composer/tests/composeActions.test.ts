import _ from "lodash";

import { compileToOldSpec, compose } from "@src/index";
import { ActionKindAST, EndpointBodyAST, EndpointCardinality } from "@src/types/ast";
import { CreateEndpointDef, EndpointType, UpdateEndpointDef } from "@src/types/definition";

describe("custom actions", () => {
  it("succeeds for basic composite create", () => {
    const bp = `
    model Org {
      field is_new { type boolean }
      field name { type string }
      reference extras { to OrgExtra, unique }
    }
    model OrgExtra {
      relation org { from Org, through extras }
    }
    model OrgOwner { reference org { to Org } }

    entrypoint Orgs {
      target Org as org
      create endpoint {
        action {
          create OrgExtra as e {}
          create org {
            set is_new true
            set extras e
          }
          create OrgOwner as oo {
            set org_id org.id
          }
        }
      }
    }
    `;
    const def = compose(compileToOldSpec(bp));
    const endpoint = def.entrypoints[0].endpoints[0] as CreateEndpointDef;
    expect(endpoint.actions).toMatchSnapshot();
  });
  it("succeeds for basic update with a deny rule", () => {
    const bp = `
    model Org {
      field name { type string }
      field description { type string }
      field uuid { type string }
      reference extras { to OrgExtra, unique }
    }
    model OrgExtra {
      relation org { from Org, through extras }
    }
    entrypoint Orgs {
      target Org as org
      update endpoint {
        action {
          update org as ox {
            set name "new name"
            deny { uuid }
          }
        }
      }
    }`;
    const def = compose(compileToOldSpec(bp));
    const endpoint = def.entrypoints[0].endpoints[0] as UpdateEndpointDef;
    expect(endpoint.actions).toMatchSnapshot();
  });
  it("succeeds with nested sibling reference", () => {
    const bp = `
    model Org {
      field name { type string }
      field name2 { type string }
      field name3 { type string }
    }
    entrypoint Org {
      target Org
      create endpoint {
        action {
          create {
            set name3 name2
            set name2 name
          }
        }
      }
    }
    `;
    const def = compose(compileToOldSpec(bp));
    const endpoint = def.entrypoints[0].endpoints[0] as CreateEndpointDef;
    expect(endpoint.actions).toMatchSnapshot();
    expect(endpoint.fieldset).toMatchSnapshot();
  });
  it("fails when reference and its field are being set at the same time", () => {
    const bp = `
    model Org { relation repos { from Repo, through org }}
    model Repo { reference org { to Org }}
    entrypoint Org {
      target Org as org
      entrypoint Repos {
        target repos as repo
        create endpoint {
          action {
            create repo {
              set org_id 1
              set org org
            }
          }
        }
      }
    }
    `;
    const spec = compileToOldSpec(bp);
    expect(() => compose(spec)).toThrowErrorMatchingInlineSnapshot(`"Found duplicates: [org_id]"`);
  });
  it("correctly sets parent context", () => {
    const bp = `
    model Org {
      relation repos { from Repo, through org }
    }
    model Repo {
      reference org { to Org }
      field name { type string }
    }
    entrypoint Orgs {
      target Org as myorg
      entrypoint Repos {
        target repos as myrepo
        create endpoint {}
      }
    }
    `;
    const def = compose(compileToOldSpec(bp));
    const endpoint = def.entrypoints[0].entrypoints[0].endpoints[0] as CreateEndpointDef;
    expect(endpoint.actions).toMatchSnapshot();
  });

  it("can create nested relations through transient references", () => {
    const bp = `
    model Org { relation repos { from Repo, through org } relation logs { from OrgLog, through org } }
    model Repo { reference org { to Org } field name { type string } }
    model OrgLog { reference org { to Org } }

    entrypoint R {
      target Repo as repo
      create endpoint {
        action {
          create repo {}
          create repo.org.logs as log {}
        }
      }
    }
    `;
    const def = compose(compileToOldSpec(bp));
    const endpoint = def.entrypoints[0].endpoints[0] as CreateEndpointDef;
    expect(endpoint.actions).toMatchSnapshot();
  });
  it("can update deeply nested references", () => {
    const bp = `
    model Org {
      field name { type string }
      relation repos { from Repo, through org }
    }
    model Repo { reference org { to Org } relation issues { from Issue, through repo } }
    model Issue { reference repo { to Repo } }

    entrypoint I {
      target Issue as issue
      update endpoint {
        action {
          update issue {}
          update issue.repo.org as org {}
        }
      }
    }
    `;
    const def = compose(compileToOldSpec(bp));
    const endpoint = def.entrypoints[0].endpoints[0] as UpdateEndpointDef;
    expect(endpoint.actions).toMatchSnapshot();
  });
  it("fails when default action override is invalid type", () => {
    const bp = `
    model Org {
      field name { type string }
    }
    entrypoint Orgs {
      target Org
      update endpoint {
        action {
          create {}
        }
      }
    }`;
    const spec = compileToOldSpec(bp);
    expect(() => compose(spec)).toThrowErrorMatchingInlineSnapshot(
      `"Mismatching context action: overriding update endpoint with a create action"`
    );
  });
  it("succeeds with custom inputs", () => {
    const bp = `
    model Org {
      field name { type string }
      field description { type string }
      field uuid { type string }
      reference extras { to OrgExtra, unique }
    }
    model OrgExtra {
      field name { type string }
      relation org { from Org, through extras }
    }
    entrypoint Orgs {
      target Org as org
      update endpoint {
        action {
          update org as ox {
            set name "new name"
            input { description { optional } }
            reference extras through name
          }
        }
      }
    }`;
    const def = compose(compileToOldSpec(bp));
    const endpoint = def.entrypoints[0].endpoints[0] as UpdateEndpointDef;
    expect(endpoint.actions).toMatchSnapshot();
  });
  it("succeeds with arithmetic expressions in setters", () => {
    const bp = `
    model Org {
      field name { type string }
      field description { type string }
      field descLength { type integer }
    }
    entrypoint Orgs {
      target Org as org
      create endpoint {
        action {
          create {
            set name "new name"
            set description concat(name, " is great")
            set descLength length(description) + 1
          }
        }
      }
    }`;
    const def = compose(compileToOldSpec(bp));
    const endpoint = def.entrypoints[0].endpoints[0] as UpdateEndpointDef;
    expect(endpoint.actions).toMatchSnapshot();
  });
  it("fails when input and reference are on the same field", () => {
    const bp = `
    model Org {
      reference extras { to OrgExtra, unique }
    }
    model OrgExtra {
      relation org { from Org, through extras }
    }
    entrypoint Orgs {
      target Org as org
      update endpoint {
        action {
          update org as ox {
            input { extras_id }
            reference extras through id
          }
        }
      }
    }`;
    const spec = compileToOldSpec(bp);
    expect(() => compose(spec)).toThrowErrorMatchingInlineSnapshot(
      `"Found duplicates: [extras_id]"`
    );
  });
  it("succeeds when virtual input is defined and referenced", () => {
    const bp = `
    model Org { field name { type string } }

    entrypoint Orgs {
      target Org as org
      create endpoint {
        action {
          create org {
            virtual input iname { type string, validate { min 4 } }
            set name concat("Mr/Mrs ", iname)
          }
        }
      }
    }
    `;

    const def = compose(compileToOldSpec(bp));
    const endpoint = def.entrypoints[0].endpoints[0] as CreateEndpointDef;
    expect(endpoint.actions).toMatchSnapshot();
  });
  it("fails when there's an input and deny for the same field", () => {
    const bp = `
    model Org {
      field name { type string }
    }
    entrypoint Orgs {
      target Org as org
      update endpoint {
        action {
          update org as ox {
            input { name }
            deny { name }
          }
        }
      }
    }
    `;
    const spec = compileToOldSpec(bp);
    expect(() => compose(spec)).toThrowErrorMatchingInlineSnapshot(`"Found duplicates: [name]"`);
  });
  it.todo("succeeds to update through unique relation");
  it("sets default action if not given", () => {
    const bp = `
    model Org {
      field name { type string }
    }
    entrypoint Orgs {
      target Org as org
      update endpoint {}
    }
    `;
    const def = compose(compileToOldSpec(bp));
    const endpoint = def.entrypoints[0].endpoints[0] as UpdateEndpointDef;
    expect(endpoint.actions).toMatchSnapshot();
  });
  it("fails when custom action doesn't have an alias", () => {
    const bp = `
    model Org {
      field name { type string }
    }
    entrypoint Orgs {
      target Org as org
      update endpoint {
        action {
          update org {}
          create Org {}
        }
      }
    }
    `;
    const spec = compileToOldSpec(bp);
    expect(() => compose(spec)).toThrowErrorMatchingInlineSnapshot(
      `"Custom action must have an alias"`
    );
  });
  test.each(["Repo", "repo", "org"])(
    "fails when action alias uses existing model or context name %s",
    (name) => {
      const bp = `
      model Org { relation repos { from Repo, through org }}
      model Repo { reference org { to Org }}
      entrypoint O {
        target Org as org
        entrypoint R {
          target repos as repo
          create endpoint {
            action {
              create repo {}
              create Repo as ${name} {}
            }
          }
        }
      }
    `;
      const spec = compileToOldSpec(bp);
      expect(() => compose(spec)).toThrowError(
        `Cannot name an action with ${name}, name already exists in the context`
      );
    }
  );
  it.todo("gives proper error when nested cycle is detected");
  // create user { create profile { create user {} } }

  it("creates actions in custom endpoint", () => {
    const bp = `
    model Org { field name { type string } }
    model Log {}

    entrypoint Orgs {
      target Org as org

      custom endpoint {
        cardinality one
        method GET
        path "customGet"
      }
      custom endpoint {
        cardinality one
        method PATCH
        path "customUpdate"

        action {
          update {}
        }
      }
      custom endpoint {
        cardinality one
        method DELETE
        path "customDelete"

        action {
          delete {}
        }
      }

      custom endpoint {
        cardinality many
        method GET
        path "customList"
      }
      custom endpoint {
        cardinality many
        method POST
        path "customCreate"

        action {
          create {}
        }
      }
    }
    `;

    const def = compose(compileToOldSpec(bp));
    const endpoints = def.entrypoints[0].endpoints;

    expect(endpoints).toMatchSnapshot();
  });

  // --- test missing/unallowed endpoint properties
  _.chain<EndpointBodyAST["kind"][]>(["cardinality", "method", "path"])
    .forEach((property) => {
      it(`fails when "${property}" property is missing in custom endpoint`, () => {
        const bp = `
          model Org { field name { type string } }
          model Log {}

          entrypoint Orgs {
            target Org as org
            custom endpoint {
              // in each iteration skip one property
              ${property === "cardinality" ? "" : "cardinality many"}
              ${property === "method" ? "" : "method POST"}
              ${property === "path" ? "" : 'path "somePath"'}

              action {
                create Log as log {}
              }
            }
          }
        `;

        expect(() => compose(compileToOldSpec(bp))).toThrowError(
          `Property "${property}" is required for custom endpoints`
        );
      });

      _.chain<EndpointType[]>(["get", "list", "create", "update", "delete"])
        .forEach((epType) => {
          it(`fails when "${property}" property is used in "${epType}" endpoint`, () => {
            const bp = `
              model Org {}

              entrypoint Orgs {
                target Org
                ${epType} endpoint {
                  // show one property in each iteration
                  ${property === "cardinality" ? "cardinality many" : ""}
                  ${property === "method" ? "method POST" : ""}
                  ${property === "path" ? 'path "somePath"' : ""}
                }
              }
            `;

            expect(() => compose(compileToOldSpec(bp))).toThrowError(
              `Property "${property}" is not allowed for "${epType}" endpoints`
            );
          });
        })
        .value();
    })
    .value();

  // --- test invalid action types in custom endpoints
  _.chain<[EndpointCardinality, ActionKindAST[]]>([
    ["one", ["create"]],
    ["many", ["update", "delete"]],
  ])
    .map(([cardinality, actions]) => {
      console.log("endpoint", cardinality);

      return _.map(actions, (a): [EndpointCardinality, ActionKindAST] => [cardinality, a]);
    })
    .flatMap()
    .forEach(([cardinality, action]) => {
      it(`fails when creating "${cardinality}" endpoint with unallowed action "${action}"`, () => {
        const bp = `
        model Org {}
        model Log {}

        entrypoint Orgs {
          target Org as org
          custom endpoint {
            cardinality ${cardinality}
            method POST
            path "somePath"

            action {
              ${action} org {}
            }
          }
        }
        `;

        expect(() => compose(compileToOldSpec(bp))).toThrowError(
          `"custom-${cardinality}" endpoint does not allow "${action}" action`
        );
      });
    })
    .value();

  it(`fails on duplicate endpoint paths`, () => {
    const bp = `
      model Org {}

      entrypoint Orgs {
        target Org as org

        custom endpoint {
          cardinality many
          method POST
          path "someDuplicatePath"
        }

        custom endpoint {
          cardinality many
          method POST
          path "someDuplicatePath"
        }
      }
      `;

    expect(() => compose(compileToOldSpec(bp))).toThrowError(
      `Custom endpoints on the same HTTP method must have unique paths in one entrypoint ("Orgs")`
    );
  });
});
