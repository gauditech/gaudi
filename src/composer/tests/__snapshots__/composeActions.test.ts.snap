// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`custom actions can create nested relations through transient references 1`] = `
[
  {
    "alias": "repo",
    "changeset": {
      "name": {
        "fieldsetAccess": [
          "name",
        ],
        "kind": "fieldset-input",
        "required": true,
        "type": "text",
      },
      "org_id": {
        "fieldsetAccess": [
          "org_id",
        ],
        "kind": "fieldset-input",
        "required": true,
        "type": "integer",
      },
    },
    "kind": "create-one",
    "model": "Repo",
    "select": [
      {
        "alias": "org",
        "kind": "reference",
        "name": "org",
        "namePath": [
          "repo",
          "org",
        ],
        "select": [
          {
            "alias": "id",
            "kind": "field",
            "name": "id",
            "namePath": [
              "repo",
              "org",
              "id",
            ],
            "refKey": "Org.id",
          },
        ],
      },
    ],
    "targetPath": [
      "repo",
    ],
  },
  {
    "alias": "log",
    "changeset": {
      "org_id": {
        "kind": "reference-value",
        "target": {
          "access": [
            "org",
            "id",
          ],
          "alias": "repo",
        },
        "type": "integer",
      },
    },
    "kind": "create-one",
    "model": "OrgLog",
    "select": [],
    "targetPath": [
      "repo",
      "org",
      "logs",
    ],
  },
]
`;

exports[`custom actions can update deeply nested references 1`] = `
[
  {
    "alias": undefined,
    "changeset": {
      "repo_id": {
        "fieldsetAccess": [
          "repo_id",
        ],
        "kind": "fieldset-input",
        "required": false,
        "type": "integer",
      },
    },
    "filter": undefined,
    "kind": "update-one",
    "model": "Issue",
    "select": [],
    "targetPath": [
      "issue",
    ],
  },
  {
    "alias": "org",
    "changeset": {
      "name": {
        "fieldsetAccess": [
          "org",
          "name",
        ],
        "kind": "fieldset-input",
        "required": false,
        "type": "text",
      },
    },
    "filter": undefined,
    "kind": "update-one",
    "model": "Org",
    "select": [],
    "targetPath": [
      "issue",
      "repo",
      "org",
    ],
  },
]
`;

exports[`custom actions correctly sets parent context 1`] = `
[
  {
    "alias": "myrepo",
    "changeset": {
      "name": {
        "fieldsetAccess": [
          "name",
        ],
        "kind": "fieldset-input",
        "required": true,
        "type": "text",
      },
      "org_id": {
        "kind": "reference-value",
        "target": {
          "access": [
            "id",
          ],
          "alias": "myorg",
        },
        "type": "integer",
      },
    },
    "kind": "create-one",
    "model": "Repo",
    "select": [],
    "targetPath": [
      "myrepo",
    ],
  },
]
`;

exports[`custom actions sets default action if not given 1`] = `
[
  {
    "alias": undefined,
    "changeset": {
      "name": {
        "fieldsetAccess": [
          "name",
        ],
        "kind": "fieldset-input",
        "required": false,
        "type": "text",
      },
    },
    "filter": undefined,
    "kind": "update-one",
    "model": "Org",
    "select": [],
    "targetPath": [
      "org",
    ],
  },
]
`;

exports[`custom actions succeeds for basic composite create 1`] = `
[
  {
    "alias": "e",
    "changeset": {},
    "kind": "create-one",
    "model": "OrgExtra",
    "select": [
      {
        "alias": "id",
        "kind": "field",
        "name": "id",
        "namePath": [
          "e",
          "id",
        ],
        "refKey": "OrgExtra.id",
      },
    ],
    "targetPath": [
      "OrgExtra",
    ],
  },
  {
    "alias": "org",
    "changeset": {
      "extras_id": {
        "kind": "reference-value",
        "target": {
          "access": [
            "id",
          ],
          "alias": "e",
        },
        "type": "integer",
      },
      "is_new": {
        "kind": "literal",
        "type": "boolean",
        "value": true,
      },
      "name": {
        "fieldsetAccess": [
          "name",
        ],
        "kind": "fieldset-input",
        "required": true,
        "type": "text",
      },
    },
    "kind": "create-one",
    "model": "Org",
    "select": [
      {
        "alias": "id",
        "kind": "field",
        "name": "id",
        "namePath": [
          "org",
          "id",
        ],
        "refKey": "Org.id",
      },
    ],
    "targetPath": [
      "org",
    ],
  },
  {
    "alias": "oo",
    "changeset": {
      "org_id": {
        "kind": "reference-value",
        "target": {
          "access": [
            "id",
          ],
          "alias": "org",
        },
        "type": "integer",
      },
    },
    "kind": "create-one",
    "model": "OrgOwner",
    "select": [],
    "targetPath": [
      "OrgOwner",
    ],
  },
]
`;

exports[`custom actions succeeds for basic update with a deny rule 1`] = `
[
  {
    "alias": "ox",
    "changeset": {
      "description": {
        "fieldsetAccess": [
          "ox",
          "description",
        ],
        "kind": "fieldset-input",
        "required": false,
        "type": "text",
      },
      "extras_id": {
        "fieldsetAccess": [
          "ox",
          "extras_id",
        ],
        "kind": "fieldset-input",
        "required": false,
        "type": "integer",
      },
      "name": {
        "kind": "literal",
        "type": "text",
        "value": "new name",
      },
    },
    "filter": undefined,
    "kind": "update-one",
    "model": "Org",
    "select": [],
    "targetPath": [
      "org",
    ],
  },
]
`;

exports[`custom actions succeeds with custom inputs 1`] = `
[
  {
    "alias": "ox",
    "changeset": {
      "description": {
        "fieldsetAccess": [
          "description",
        ],
        "kind": "fieldset-input",
        "required": false,
        "type": "text",
      },
      "extras": {
        "fieldsetAccess": [
          "name",
        ],
        "kind": "fieldset-reference-input",
        "throughField": {
          "name": "name",
          "refKey": "OrgExtra.name",
        },
      },
      "name": {
        "kind": "literal",
        "type": "text",
        "value": "new name",
      },
      "uuid": {
        "fieldsetAccess": [
          "ox",
          "uuid",
        ],
        "kind": "fieldset-input",
        "required": false,
        "type": "text",
      },
    },
    "filter": undefined,
    "kind": "update-one",
    "model": "Org",
    "select": [],
    "targetPath": [
      "org",
    ],
  },
]
`;
